File: Code/Components/DanpheEMR.ServerModel/BillingModels/InvoicePrintVMs.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel.BillingModels
{

    public class BilPrint_PatientInfoVM
    {
        public int PatientId { get; set; }
        public string PatientCode { get; set; }
        public string ShortName { get; set; }
        public string Gender { get; set; }
        public DateTime DateOfBirth { get; set; }
        public string Age { get; set; }

        public int CountryId { get; set; }
        public string CountryName { get; set; }
        public int CountrySubDivisionId { get; set; }
        public string CountrySubDivisionName { get; set; }
        public string Address { get; set; }
        public string PhoneNumber { get; set; }
        public int MembershipTypeId { get; set; }
        public string MunicipalityName { get; set; }
        public string MembershipTypeName { get; set; }
        public string PANNumber { get; set; }
        public string PatientNameLocal { get; set; }
        public string Ins_NshiNumber { get; set; }

        //since we're getting datatable, we need to map it to the PatientInfoVM by First Serialize and the DeSerialize.
        public static BilPrint_PatientInfoVM MapDataTableToSingleObject(DataTable patInfo)
        {
            BilPrint_PatientInfoVM retObj = new BilPrint_PatientInfoVM();
            if (patInfo != null)
            {
                string strPatData = JsonConvert.SerializeObject(patInfo);
                //Datatable contains array, we need to deserialize into list then take the first one.
                List<BilPrint_PatientInfoVM> patList = JsonConvert.DeserializeObject<List<BilPrint_PatientInfoVM>>(strPatData);
                if (patList != null && patList.Count > 0)
                {
                    retObj = patList.First();
                }
            }
            return retObj;
        }

    }

    public class BilPrint_InvoiceInfoVM
    {
        public int InvoiceNumber { get; set; }
        public string InvoiceCode { get; set; }
        public string InvoiceNumFormatted { get; set; }
        public DateTime TransactionDate { get; set; }
        public int FiscalYearId { get; set; }
        public string FiscalYear { get; set; }

        public string PaymentMode { get; set; }
        public string BillStatus { get; set; }
        public string TransactionType { get; set; }
        public string InvoiceType { get; set; }
        public int PrintCount { get; set; }
        public double SubTotal { get; set; }
        public double DiscountAmount { get; set; }
        public double TaxableAmount { get; set; }
        public double NonTaxableAmount { get; set; }
        public double TotalAmount { get; set; }

        public int BillingTransactionId { get; set; }
        public DateTime? PaidDate { get; set; }
        public double? Tender { get; set; }
        public double? Change { get; set; }
        public string Remarks { get; set; }
        public bool IsInsuranceBilling { get; set; }
        public Int64? ClaimCode { get; set; }
        public int? CrOrganizationId { get; set; }
        public string CrOrganizationName { get; set; }
        public string UserName { get; set; }
        public int CounterId { get; set; }
        public string CounterName { get; set; }
        public string LabTypeName { get; set; }

        public int? PackageId { get; set; }
        public string PackageName { get; set; }
        public double DepositAvailable { get; set; }
        public double DepositUsed { get; set; }
        public double DepositReturnAmount { get; set; }
        public double DepositBalance { get; set; }

        //since we're getting datatable, we need to map it to the InvoiceInfoVM by First Serialize and the DeSerialize.
        public static BilPrint_InvoiceInfoVM MapDataTableToSingleObject(DataTable invInfo)
        {
            BilPrint_InvoiceInfoVM retObj = new BilPrint_InvoiceInfoVM();
            if (invInfo != null)
            {
                string strInvData = JsonConvert.SerializeObject(invInfo);
                //Datatable contains array, we need to deserialize into list then take the first one.
                List<BilPrint_InvoiceInfoVM> invList = JsonConvert.DeserializeObject<List<BilPrint_InvoiceInfoVM>>(strInvData);
                if (invList != null && invList.Count > 0)
                {
                    retObj = invList.First();
                }
            }
            return retObj;
        }

    }

    public class BilPrint_InvoiceItemVM
    {
        public int BillingTransactionItemId { get; set; }
        public int ServiceDepartmentId { get; set; }
        public int ItemId { get; set; }
        public string ServiceDepartmentName { get; set; }
        public string ItemName { get; set; }
        public double Price { get; set; }
        public double Quantity { get; set; }
        public double SubTotal { get; set; }
        public double DiscountAmount { get; set; }
        public double TotalAmount { get; set; }
        public int? ProviderId { get; set; }
        public int? RequestedBy { get; set; }
        public string ProviderName { get; set; }
        public string RequestedByName { get; set; }
        public string PriceCategory { get; set; }


        //Map Datatable to object by first serializing then deserializing into required format.
        //Incoming datatable shoul have exact same column/datatype as that of the Object to be Mapped.
        public static List<BilPrint_InvoiceItemVM> MapDataTableToObjectList(DataTable dtInvoiceItem)
        {
            List<BilPrint_InvoiceItemVM> retListObj = new List<BilPrint_InvoiceItemVM>();
            if (dtInvoiceItem != null)
            {
                string strInvItms = JsonConvert.SerializeObject(dtInvoiceItem);
                retListObj = JsonConvert.DeserializeObject<List<BilPrint_InvoiceItemVM>>(strInvItms);
            }
            return retListObj;
        }
    }

    public class BilPrint_VisitInfoVM
    {
        public int PatientVisitId { get; set; }
        public string VisitCode { get; set; }
        public DateTime? AdmissionDate { get; set; }
        public DateTime? DischargeDate { get; set; }
        public string WardName { get; set; }
        public int? BedNumber { get; set; }
        public string BedCode { get; set; }
        public string ConsultingDoctor { get; set; }
        public int? ConsultingDoctorId { get; set; }
        public List<int?> ItemsRequestingDoctorsId { get; set; }
        public string ItemsRequestingDoctors { get; set; }

        //since we're getting datatable, we need to first Serialize then again Deserialize to get a proper object.
        public static BilPrint_VisitInfoVM MapDataTableToSingleObject(DataTable dtVisInfo)
        {
            BilPrint_VisitInfoVM retObj = new BilPrint_VisitInfoVM();
            if (dtVisInfo != null)
            {
                string strVisData = JsonConvert.SerializeObject(dtVisInfo);
                //Datatable contains array, we need to deserialize into list then take the first one.
                List<BilPrint_VisitInfoVM> invList = JsonConvert.DeserializeObject<List<BilPrint_VisitInfoVM>>(strVisData);
                if (invList != null && invList.Count > 0)
                {
                    retObj = invList.First();
                }
            }
            return retObj;
        }

    }

    public class BilPrint_DepositListVM
    {
        public int DepositId { get; set; }
        public int ReceiptNo { get; set; }
        public string FiscalYearFormatted { get; set; }
        public string DepositReceiptNoFormattted { get; set; }
        public string DepositType { get; set; }
        public double Amount { get; set; }
        public DateTime CreatedOn { get; set; }
        public string UserName { get; set; }

        //Map Datatable to object by first serializing then deserializing into required format.
        //Incoming datatable shoul have exact same column/datatype as that of the Object to be Mapped.
        public static List<BilPrint_DepositListVM> MapDataTableToObjectList(DataTable dtDeposit)
        {
            List<BilPrint_DepositListVM> retListObj = new List<BilPrint_DepositListVM>();
            if (dtDeposit != null)
            {
                string strDepItms = JsonConvert.SerializeObject(dtDeposit);
                retListObj = JsonConvert.DeserializeObject<List<BilPrint_DepositListVM>>(strDepItms);
            }
            return retListObj;
        }
    }

}

File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/billing/shared/invoice-print-vms.ts
import { PharmacyBillItemVM } from "./discharge-bill.view.model";

export class BilPrint_VM {
  PatientInfo: BilPrint_PatientInfoVM = null;
  InvoiceInfo: BilPrint_InvoiceInfoVM = null;
  InvoiceItems: Array<BilPrint_InvoiceItemVM> = null;
  VisitInfo: BilPrint_VisitInfoVM = null;
  DepositList: Array<BilPrint_DepositListVM> = null;
  IsInvoiceFound: boolean = false;
  //!Krishna, 31stMarch'23 Copied from EMR_V2.3.1_manipal for Discharge Statement
  PharmacyInvoiceItems: Array<PharmacyBillItemVM> = null
  DischargeInfo: BilPrint_DischargeInfoVM = null;
  BillingInvoiceSummary: Array<BilPrintBillingSummaryVM> = null;
  PharmacySummary: Array<BilPrintPharmacySummaryVM> = null;

}

export class SSFBil_VM {
  PatientInfo = new Array<SSF_Invoice_PatInfo>();
  InvoiceInfo = new Array<SSF_InvoiceInfoVM>();
  InvoiceItems = new Array<SSF_InvoiceItems>();
  PhrmInvoices = new Array<SSF_PhrmInvoices>();
  BillingInvoiceInfo = new Array<SSF_BillingInvoiceInfoVM>();
  BillingInvoiceItems = new Array<SSF_BillingInvoiceItems>();
  PhrmInvoiceItems = new Array<SSF_PhrmInvoiceItems>();
  LabReportInfo = new Array<SSF_LabReportInfo>();
  RadiologyReportInfo = new Array<SSF_RadiologyReportInfo>();
  BillingInvoiceReturns = new Array<SSFInvoiceReturnsList>();
  PharmacyInvoiceReturns = new Array<SSFInvoiceReturnsList>();
}

export class SSF_LabReportInfo {
  public PatientId: number = 0;
  public ClaimCode: number = 0;
  public RequisitionIdCSV: string = "";
}

export class SSF_RadiologyReportInfo {
  public PatientId: number = 0;
  public ClaimCode: number = 0;
  public RequisitionIdCSV: number = 0;
}


export class SSF_PhrmInvoiceItems {
  public UnitPrice: number = 0;
  public ServiceCode: string = "ADJ02" //* This is hardcoded and not used anywhere else.
  public PatientId: number = 0;
  public Quantity: number = 0;
  public ClaimCode: number = 0;
  public InvoiceId: number = null;
  public ItemName: string = null;
}

export class SSF_BillingInvoiceItems {
  public BillingTransactionItemId: number = null;
  public PatientId: number = null;
  public BillingTransactionId: number = null;
  public ServiceDepartmentId: number = null;
  public ItemId: number = null;
  public ServiceDepartmentName: string = null;
  public ItemName: string = null;
  public Price: number = null;
  public Quantity: number = null;
  public SubTotal: number = null;
  public DiscountAmount: number = null;
  public TotalAmount: number = null;
  public PerformerId: number = null;
  public PerformerName: string = null;
  public RequestedById: number = null;
  public RequestedByName: string = null;
  public RequestedBy: string = null;
  public PriceCategory: string = null;
  public ServiceCode: string = null;
  public ItemCode: string = null;
  public IsCoPayment: boolean = false;
  public DiscountPercent: number = null;
}

export class SSF_BillingInvoiceInfoVM {
  public PatientId: number = 0;
  public NetReceivableAmount: number = 0;
  public InvoiceNumber: number = null;
  public InvoiceCode: string = null;
  public InvoiceNumFormatted: string = null;
  public TransactionDate: string = null;
  public FiscalYearId: number = null;
  public FiscalYear: string = null;
  public PaymentMode: string = null;
  public PaymentDetails: string = null;
  public BillStatus: string = null;
  public TransactionType: string = null;
  public InvoiceType: string = null;
  public PrintCount: number = null;
  public SubTotal: number = null;
  public DiscountAmount: number = null;
  public TaxableAmount: number = null;
  public NonTaxableAmount: number = null;
  public TotalAmount: number = null;
  public BillingTransactionId: number = null;
  public PaidDate: string = null;
  public Tender: number = null;
  public Change: number = null;
  public Remarks: string = null;
  public IsInsuranceBilling: boolean = null;
  public ClaimCode: number = null;//sud:1-oct'21: Changed datatype from String to Number in all places
  public CrOrganizationId: number = null;
  public CrOrganizationName: string = null;
  public UserName: string = null;
  public CounterId: number = null;
  public CounterName: string = null;
  public LabTypeName: string = null;
  public PackageId: number = null;
  public PackageName: string = null;
  public DepositAvailable: number = null;
  public DepositUsed: number = null;
  public DepositReturnAmount: number = null;
  public DepositBalance: number = null;
  public TaxTotal: number = null;
  public ReceivedAmount: number = null;
  public InvoiceDate: number = null;
}

// export class SSF_PhrmInvoices {
//   public UnitPrice: number = 0;
//   public ServiceCode: string = "ADJ02" //* This is hardcoded and not used anywhere else.
//   public PatientId: number = 0;
//   public Quantity: number = 0;
//   public ClaimCode: number = 0;
//   public InvoiceId: number = null;
//   public InvoiceNo: number = 0;
//   public InvoiceDate: string = "";
// }

export class SSF_PhrmInvoices {
  public InvoiceNumber: number = 0;
  public InvoiceNoFormatted: string = "";
  public TotalAmount: number = 0;
  public ReceivedAmount: number = 0;
  public BalanceAmount: number = 0;
  public ClaimCode: number = 0;
  public PatientId: number = 0;
  public InvoiceId: number = null;
  public InvoiceDate: string = "";
}

export class SSF_Invoice_PatInfo {
  public PolicyNo: string = null;
  public PatientId: number = null;
  public PatientCode: string = null;
  public ShortName: string = null;
  public Gender: string = null;
  public DateOfBirth: string = null;
  public Age: string = null;
  public CountryId: number = null;
  public CountryName: string = null;
  public CountrySubDivisionId: number = null;
  public CountrySubDivisionName: string = null;
  public Address: string = null;
  public PhoneNumber: string = null;
  public MembershipTypeId: number = null;
  public MembershipTypeName: string = null;
  public PANNumber: string = null;
  public PatientNameLocal: string = null;
  public Ins_NshiNumber: string = null;
  public MunicipalityName: string = null;
  public PolicyHolderUID: string = null;
  public PolicyHolderEmployerId: string = null;
  public SchemeType: number = null;
  public Admitted: string = "0"; //* server sends this as string hence cannot make it boolean.
  public Diagnosis: string = null;
  public AdmissionDate: string = null;
  public DischargeDate: string = null;
  public DischargeTypeName: string = null;
  public CaseSummary: string = null;
  public IsDead: number = null;
  public ClaimCode: number = 0;
  public VisitType: string = "";
  public VisitCreationDate: string = "";
  public WardNumber: string = null;
  public SchemeName: string = null;
  public DepartmentName: string = null;
}

export class SSFClaimList {
  public IsSelected: boolean = false;
  public InvoiceNo: number;
  //public ClaimCode : number;
  public SSFClaimCode: string;
  public InvoiceTotalAmount: number;
  public Cash: number;
  public Credit: number;
  public ClaimedAmount: number;
  public Status: boolean;
  public BillingTransactionId: number;
  public InvoiceNoFormatted: string = "";
  public BookingStatus: string = "";
  public ModuleName: string = "";
}

export class PatientWiseSSFClaimList {
  public IsSelected: boolean = false;
  public PolicyNo: string;
  public PatientName: string;
  public PatientCode: string;
  public Address: string;
  public PatientId: number = 0;
  public ClaimCode: number = 0;
  public InvoiceList: Array<SSFClaimList> = new Array<SSFClaimList>();
  public InvoiceReturnsList = new Array<SSFInvoiceReturnsList>();
  public SchemeType: number = null;
}

export class SSF_InvoiceItems {
  public BillingTransactionItemId: number = null;
  public PatientId: number = null;
  public BillingTransactionId: number = null;
  public ServiceDepartmentId: number = null;
  public ItemId: number = null;
  public ServiceDepartmentName: string = null;
  public ItemName: string = null;
  public Price: number = null;
  public Quantity: number = null;
  public SubTotal: number = null;
  public DiscountAmount: number = null;
  public TotalAmount: number = null;
  public PerformerId: number = null;
  public PerformerName: string = null;
  public RequestedById: number = null;
  public RequestedByName: string = null;
  public RequestedBy: string = null;
  public PriceCategory: string = null;
  public ServiceCode: string = null;
}

export class BilPrint_PatientInfoVM {
  public PatientId: number = null;
  public PatientCode: string = null;
  public ShortName: string = null;
  public Gender: string = null;
  public DateOfBirth: string = null;
  public Age: string = null;
  public CountryId: number = null;
  public CountryName: string = null;
  public CountrySubDivisionId: number = null;
  public CountrySubDivisionName: string = null;
  public Address: number = null;
  public WardNumber: number = null;
  public PhoneNumber: string = null;
  public MembershipTypeId: number = null;
  public MembershipTypeName: string = null;
  public PANNumber: string = null;
  public PatientNameLocal: string = null;
  public Ins_NshiNumber: string = null;
  public MunicipalityName: string = null;
  public SSFPolicyNo: string = "";
  public PolicyNo: string = "";
}

export class BilPrint_InvoiceInfoVM {
  public InvoiceNumber: number = null;
  public InvoiceCode: string = null;
  public InvoiceNumFormatted: string = null;
  public TransactionDate: string = null;
  public FiscalYearId: number = null;
  public FiscalYear: string = null;
  public PaymentMode: string = null;
  public PaymentDetails: string = null;
  public BillStatus: string = null;
  public TransactionType: string = null;
  public InvoiceType: string = null;
  public PrintCount: number = null;
  public SubTotal: number = null;
  public DiscountAmount: number = null;
  public TaxableAmount: number = null;
  public NonTaxableAmount: number = null;
  public TotalAmount: number = null;
  public BillingTransactionId: number = null;
  public PaidDate: string = null;
  public Tender: number = null;
  public Change: number = null;
  public Remarks: string = null;
  public IsInsuranceBilling: boolean = null;
  public ClaimCode: number = null;//sud:1-oct'21: Changed datatype from String to Number in all places
  public CrOrganizationId: number = null;
  public CreditOrganizationName: string = null;
  public UserName: string = null;
  public CounterId: number = null;
  public CounterName: string = null;
  public LabTypeName: string = null;
  public PackageId: number = null;
  public PackageName: string = null;
  public DepositAvailable: number = null;
  public DepositUsed: number = null;
  public DepositReturnAmount: number = null;
  public DepositBalance: number = null;
  public TaxTotal: number = null;
  public ReceivedAmount: number = null;
  public SchemeName: string = null;
  public OtherCurrencyDetail: string = null;
}


export class BilPrint_InvoiceItemVM {
  public BillingTransactionItemId: number = null;
  public ServiceDepartmentId: number = null;
  public IntegrationItemId: number = null;
  public ItemId: number = null;
  public ItemCode: string = null;
  public ServiceDepartmentName: string = null;
  public ItemName: string = null;
  public Price: number = null;
  public Quantity: number = null;
  public SubTotal: number = null;
  public DiscountPercent: number = 0;
  public DiscountAmount: number = null;
  public TotalAmount: number = null;
  public PerformerId: number = null;
  public PerformerName: string = null;
  public RequestedById: number = null;
  public RequestedByName: string = null;
  public RequestedBy: string = null;
  public PriceCategory: string = null;
  public BillDate: string = null;
  public IsCoPayment: boolean = false;
  public ServiceCategoryCode: string = "";
  public ServiceCategoryName: string = "";
}


export class BilPrint_VisitInfoVM {
  public PatientVisitId: number = null;
  public VisitCode: string = null;
  public AdmissionDate: string = null;
  public DischargeDate: string = null;
  public WardName: string = null;
  public BedNumber: string = null;
  public BedCode: string = null;
  public ConsultingDoctor: string = null;
  public ItemsRequestingDoctorsId: Array<number> = new Array<number>();
  public ItemsRequestingDoctors: string = null;
  public DepartmentName: string = null;
}

export class BilPrint_DepositListVM {
  public DepositId: number = null;
  public ReceiptNo: number = null;
  public FiscalYearFormatted: string = null;
  public DepositReceiptNoFormattted: string = null;
  // public DepositType: string = null;
  public TransactionType: string = null;
  public InAmount: number = 0;
  public OutAmount: number = 0;
  public CreatedOn: string = null;
  public UserName: string = null;
}

export class SSF_InvoiceInfoVM {
  public PatientId: number = 0;
  public NetReceivableAmount: number = 0;
  public InvoiceNumber: number = null;
  public InvoiceCode: string = null;
  public InvoiceNumFormatted: string = null;
  public TransactionDate: string = null;
  public FiscalYearId: number = null;
  public FiscalYear: string = null;
  public PaymentMode: string = null;
  public PaymentDetails: string = null;
  public BillStatus: string = null;
  public TransactionType: string = null;
  public InvoiceType: string = null;
  public PrintCount: number = null;
  public SubTotal: number = null;
  public DiscountAmount: number = null;
  public TaxableAmount: number = null;
  public NonTaxableAmount: number = null;
  public TotalAmount: number = null;
  public BillingTransactionId: number = null;
  public PaidDate: string = null;
  public Tender: number = null;
  public Change: number = null;
  public Remarks: string = null;
  public IsInsuranceBilling: boolean = null;
  public ClaimCode: number = null;//sud:1-oct'21: Changed datatype from String to Number in all places
  public CrOrganizationId: number = null;
  public CreditOrganizationName: string = null;
  public UserName: string = null;
  public CounterId: number = null;
  public CounterName: string = null;
  public LabTypeName: string = null;
  public PackageId: number = null;
  public PackageName: string = null;
  public DepositAvailable: number = null;
  public DepositUsed: number = null;
  public DepositReturnAmount: number = null;
  public DepositBalance: number = null;
  public TaxTotal: number = null;
  public ReceivedAmount: number = null;
  public InvoiceDate: number = null;
}

export class BilPrint_DischargeInfoVM {
  public DischargeStatementId: number;
  public StatementNo: string;
  public StatementTime: string;
  public StatementDate: string;
}

export class BilPrintBillingSummaryVM {
  public GroupName: string;
  public SubTotal: number;
  public DiscountAmount: number;
  public TotalAmount: number;
}

export class BilPrintPharmacySummaryVM {
  public GroupName: string;
  public SubTotal: number;
  public DiscountAmount: number;
  public TotalAmount: number;
}

export class SSFInvoiceReturnsList {
  public PatientId: number = null;
  public ReturnId: number = null;
  public CreditNoteNumber: number = null;
  public CreditNoteNumberFormatted: string = "";
  public ClaimCode: number = null;
  public TotalAmount: number = 0;
  public ModuleName: string = "";
}
